src/script/sign.cpp

```diff
@@ -185,6 +263,200 @@ bool ProduceSignature(const BaseSignatureCreator& creator, const CScript& fromPu
+                        P2SH = true;
+                        {
+                            auto redeem_script_it = psbt.redeem_scripts.find(uint160(vSolutions[0]));
+                            if (redeem_script_it != psbt.redeem_scripts.end()) {
+                                spk = redeemscript = redeem_script_it->second;
+                            } else {
+                                break;
+                            }
+                        }
+                        loop = true;
+                        break;
+                    case TX_MULTISIG: {
+                        script_ret.push_back(valtype()); // workaround CHECKMULTISIG bug
+                        SignatureData dummy_sigdata;
+                        dummy_sigdata.scriptSig = spk;
+                        Stacks redeemscript_stack(dummy_sigdata);
+                        unsigned int nSigsRequired = redeemscript_stack.script.front()[0];
+                        unsigned int nSigsHave = 0;
+                        unsigned int nPubKeys = redeemscript_stack.script.size()-2;
+                        for (unsigned int j = 0; j < nPubKeys && nSigsHave < nSigsRequired; ++j)
+                        {
+                            auto sig_it = psbt.inputs[i].partial_sigs.find(CPubKey(redeemscript_stack.script[j + 1]));
+                            if (sig_it != psbt.inputs[i].partial_sigs.end()) {
+                                script_ret.push_back(sig_it->second);
+                                ++nSigsHave;
+                            }
+                        }
+                        break;
+                    }
+                    case TX_WITNESS_V0_SCRIPTHASH:
+                        witness = true;
+                        WSH = true;
+                        {
+                            auto witness_script_it = psbt.witness_scripts.find(uint256(vSolutions[0]));
+                            if (witness_script_it != psbt.witness_scripts.end()) {
+                                spk = witnessscript = witness_script_it->second;
+                            } else {
+                                break;
+                            }
+                        }
+                        loop = true;
+                        break;
+
+                    case TX_NONSTANDARD:
+                    case TX_NULL_DATA:
+                    default:
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (witness) {
+            if (WSH) {
+                script_ret.push_back(std::vector<unsigned char>(witnessscript.begin(), witnessscript.end()));
+            }
+            sigdata.scriptWitness.stack = script_ret;
+            script_ret.clear();
+        }
+        if (P2SH) {
+            script_ret.push_back(std::vector<unsigned char>(redeemscript.begin(), redeemscript.end()));
+        }
+        sigdata.scriptSig = PushAll(script_ret);
+
+        // Test solution
```

Comment:
Perhaps you can do this test at the beginning of this function, and if succesful, delete all psbt fields? That way you can have 'dumb' combiners that work fine when combining an actually-complete version with an older version.